sklearn/tree/_tree.pyx

```diff
@@ -290,17 +291,34 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
 
 
 # Best first builder ----------------------------------------------------------
+cdef struct FrontierRecord:
+    # Record of information of a Node, the frontier for a split. Those records are
+    # maintained in a heap to access the Node with the best improvement in impurity,
+    # allowing growing trees greedily on this improvement.
+    SIZE_t node_id
+    SIZE_t start
+    SIZE_t end
+    SIZE_t pos
+    SIZE_t depth
+    bint is_leaf
+    double impurity
+    double impurity_left
+    double impurity_right
+    double improvement
 
-cdef inline int _add_to_frontier(PriorityHeapRecord* rec,
-                                 PriorityHeap frontier) nogil except -1:
-    """Adds record ``rec`` to the priority queue ``frontier``
+cdef inline bool _compare_records(
+    const FrontierRecord& left,
+    const FrontierRecord& right,
+):
+    return left.improvement < right.improvement
 
-    Returns -1 in case of failure to allocate memory (and raise MemoryError)
-    or 0 otherwise.
-    """
-    return frontier.push(rec.node_id, rec.start, rec.end, rec.pos, rec.depth,
-                         rec.is_leaf, rec.improvement, rec.impurity,
-                         rec.impurity_left, rec.impurity_right)
+cdef inline void _add_to_frontier(
+    FrontierRecord rec,
+    vector[FrontierRecord]& frontier,
+) nogil:
+    """Adds record `rec` to the priority queue `frontier`."""
+    frontier.push_back(rec)
+    push_heap(frontier.begin(), frontier.end(), &_compare_records)
```

Comment:
Isn't there a return code that should be inspected to possibly raise a `MemoryError` here?

Reply:
The `MemoryError` is raised a line above with `froniter.push_back`. This happens because the Cython declaration for `push_back` has an `except +`:

https://github.com/cython/cython/blob/756b795532151f0a8bda7960fd91f4c663385ecf/Cython/Includes/libcpp/vector.pxd#L74

According to the [Cython docs](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions), this will convert the C++ exception into a Python and raise it.

Reply:
IIRC, I think the associated interface needs to have the `except +` qualification as well to transfer the exception to Python but this is incompatible with having it `nogil` qualified.

Reply:
I think it still works. In a notebook with `%load_ext Cython`:

```python
%%cython --cplus
from libcpp.vector cimport vector

cdef void out_of_range() nogil:
    cdef vector[int] a
    
    # Raise out_of_range exception
    a.at(20)
    
def py_out_of_range():
    with nogil:
        out_of_range()
```

Calling `py_out_of_range` still raises an `IndexError`.